astropy/table/operations.py

```diff
@@ -705,116 +824,66 @@ def _join(left, right, keys=None, join_type='inner',
-
-    # Input validation
-    if join_type not in ('inner', 'exact', 'outer'):
-        raise ValueError("`join_type` arg must be one of 'inner', 'exact' or 'outer'")
-
     # Trivial case of one input array
-    if len(arrays) == 1:
-        return arrays[0]
-
-    for arr in arrays:
-        if arr.has_mixin_columns:
-            raise NotImplementedError('vstack not available for tables with mixin columns')
-
-    # Start by assuming an outer match where all names go to output
-    names = set(itertools.chain(*[arr.colnames for arr in arrays]))
-    col_name_map = get_col_name_map(arrays, names)
-
-    # If require_match is True then the output must have exactly the same
-    # number of columns as each input array
-    if join_type == 'exact':
-        for names in six.itervalues(col_name_map):
-            if any(x is None for x in names):
-                raise TableMergeError('Inconsistent columns in input arrays '
-                                      "(use 'inner' or 'outer' join_type to "
-                                      "allow non-matching columns)")
-        join_type = 'outer'
-
-    # For an inner join, keep only columns where all input arrays have that column
-    if join_type == 'inner':
-        col_name_map = OrderedDict((name, in_names) for name, in_names in six.iteritems(col_name_map)
-                                   if all(x is not None for x in in_names))
-        if len(col_name_map) == 0:
-            raise TableMergeError('Input arrays have no columns in common')
+    if len(tables) == 1:
+        return tables[0]
 
     # If there are any output columns where one or more input arrays are missing
     # then the output must be masked.  If any input arrays are masked then
     # output is masked.
-    masked = any(getattr(arr, 'masked', False) for arr in arrays)
+    masked = any(getattr(table, 'masked', False) for table in tables)
     for names in six.itervalues(col_name_map):
         if any(x is None for x in names):
             masked = True
             break
 
-    lens = [len(arr) for arr in arrays]
+    lens = [len(table) for table in tables]
     n_rows = sum(lens)
-    out = _get_out_class(arrays)(masked=masked)
-    out_descrs = get_descrs(arrays, col_name_map)
-    for out_descr in out_descrs:
-        name = out_descr[0]
-        dtype = out_descr[1:]
+    out = _get_out_class(tables)(masked=masked)
+    col_descrs = get_descrs(tables, col_name_map)
+    cols = []
+    for (name, dtype, shape), out_attrs in zip(col_descrs, col_attrs_map.values()):
+        kwargs = dict(name=name, dtype=dtype, shape=shape, length=n_rows)
         if masked:
-            out[name] = ma.array(data=np.zeros(n_rows, dtype),
-                                 mask=np.ones(n_rows, ma.make_mask_descr(dtype)))
-        else:
-            out[name] = np.empty(n_rows, dtype=dtype)
+            kwargs['mask'] = True
+
+        kwargs.update(**out_attrs)
+        cols.append(out.ColumnClass(**kwargs))
```

Comment:
@taldcroft - the key difference is here: instead of creating an empty array that becomes a column without attributes, I create an empty base column with all the attributes already set (this will be `Column` even for `QTable`, but is then converted to `Quantity` by `addcolumns` below).